lib/utils/proxy/proxy.go

```diff
@@ -58,20 +82,55 @@ type Dialer interface {
 	DialTimeout(network, address string, timeout time.Duration) (net.Conn, error)
 }
 
-type directDial struct{}
+type directDial struct {
+	useTLS bool
+}
+
+// isALPNUnsupportedError allows to determine if a teleport proxy client tried to connect to ALPN SNI proxy with raw
+// ssh protocol. After detecting this case the client fallback to TLS dialer.
+func isALPNUnsupportedError(err error) bool {
+	return err.Error() == "ssh: handshake failed: EOF"
```

Comment:
Like mentioned somewhere above, I'm curious if you've tried to find a more reliable way to determine whether ALPN is enabled or not than string matching the error message (which can legitimately happen for other reasons).

Reply:
Reworked to proxy ping call to check if proxy support ALPN